File: python/test_crypto.py
"""Test suite for all Python cryptography examples."""

import unittest
from aes_gcm_example import encrypt, decrypt, generate_key
from aes_key_wrap_example import wrap, unwrap
from ed25519_example import sign, verify
from hashing_example import sha256
from hmac_example import compute, verify as hmac_verify
from pbkdf2_example import derive, verify as pbkdf2_verify
from rsa_oaep_pss_example import encrypt as rsa_encrypt, decrypt as rsa_decrypt, sign as rsa_sign, verify as rsa_verify, generate_private_key
from x25519_hkdf_example import derive_key
from crypto_support import random_bytes, constant_time_equal


class TestCryptography(unittest.TestCase):
    def test_hashing(self):
        data = b"test data"
        digest = sha256(data)
        self.assertEqual(len(digest), 32)
        self.assertNotEqual(digest, sha256(data + b" "))

    def test_hmac(self):
        key = random_bytes(32)
        message = b"test message"
        tag = compute(key, message)
        self.assertTrue(hmac_verify(key, message, tag))
        self.assertFalse(hmac_verify(key, message + b"!", tag))

    def test_pbkdf2(self):
        password = b"test password"
        stored = derive(password)
        self.assertTrue(pbkdf2_verify(password, stored))
        self.assertFalse(pbkdf2_verify(b"wrong password", stored))

    def test_aes_gcm(self):
        key = generate_key()
        plaintext = b"secret message"
        aad = b"aad"
        envelope = encrypt(key, plaintext, aad)
        decrypted = decrypt(key, envelope, aad)
        self.assertEqual(plaintext, decrypted)

    def test_rsa_oaep_pss(self):
        private_key = generate_private_key()
        public_key = private_key.public_key()
        message = b"rsa message"
        ciphertext = rsa_encrypt(public_key, message)
        self.assertEqual(message, rsa_decrypt(private_key, ciphertext))
        signature = rsa_sign(private_key, message)
        self.assertTrue(rsa_verify(public_key, message, signature))

    def test_ed25519(self):
        from cryptography.hazmat.primitives.asymmetric import ed25519
        private_key = ed25519.Ed25519PrivateKey.generate()
        public_key = private_key.public_key()
        message = b"ed25519 message"
        signature = sign(private_key, message)
        self.assertTrue(verify(public_key, message, signature))

    def test_x25519_hkdf(self):
        from cryptography.hazmat.primitives.asymmetric import x25519
        alice = x25519.X25519PrivateKey.generate()
        bob = x25519.X25519PrivateKey.generate()
        salt = random_bytes(16)
        info = b"info"
        alice_key = derive_key(alice, bob.public_key(), salt, info)
        bob_key = derive_key(bob, alice.public_key(), salt, info)
        self.assertTrue(constant_time_equal(alice_key, bob_key))

    def test_aes_key_wrap(self):
        kek = random_bytes(32)
        content_key = random_bytes(32)
        wrapped = wrap(kek, content_key)
        recovered = unwrap(kek, wrapped)
        self.assertTrue(constant_time_equal(content_key, recovered))


if __name__ == "__main__":
    unittest.main()

File: java/src/main/java/org/example/crypto/TestCrypto.java
package org.example.crypto;

import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.util.Arrays;

public final class TestCrypto {
    public static void main(String[] args) throws Exception {
        // Hashing
        byte[] digest = HashingExample.sha256("test data".getBytes(StandardCharsets.UTF_8));
        assert digest.length == 32;

        // HMAC
        byte[] key = CryptoSupport.randomBytes(32);
        byte[] message = "test message".getBytes(StandardCharsets.UTF_8);
        byte[] tag = HmacExample.compute(key, message);
        assert HmacExample.verify(key, message, tag);
        assert !HmacExample.verify(key, (message + "!").getBytes(StandardCharsets.UTF_8), tag);

        // PBKDF2
        char[] password = "test password".toCharArray();
        Pbkdf2Example.DerivedKey stored = Pbkdf2Example.derive(password);
        assert Pbkdf2Example.verify(password, stored);
        assert !Pbkdf2Example.verify("wrong password".toCharArray(), stored);

        // AES-GCM
        javax.crypto.SecretKey aesKey = AesGcmExample.generateKey();
        byte[] plaintext = "test message".getBytes(StandardCharsets.UTF_8);
        byte[] aad = "aad".getBytes(StandardCharsets.UTF_8);
        AesGcmExample.Envelope envelope = AesGcmExample.encrypt(aesKey, plaintext, aad);
        assert Arrays.equals(plaintext, AesGcmExample.decrypt(aesKey, envelope, aad));

        // RSA-OAEP-PSS
        KeyPair keys = RsaOaepPssExample.generateKeyPair();
        byte[] rsaMsg = "rsa message".getBytes(StandardCharsets.UTF_8);
        byte[] ciphertext = RsaOaepPssExample.encrypt(keys.getPublic(), rsaMsg);
        assert Arrays.equals(rsaMsg, RsaOaepPssExample.decrypt(keys.getPrivate(), ciphertext));
        byte[] signature = RsaOaepPssExample.sign(keys.getPrivate(), rsaMsg);
        assert RsaOaepPssExample.verify(keys.getPublic(), rsaMsg, signature);

        // Ed25519
        KeyPair edKeys = Ed25519Example.generateKeyPair();
        byte[] edMsg = "ed25519 message".getBytes(StandardCharsets.UTF_8);
        byte[] edSig = Ed25519Example.sign(edKeys.getPrivate(), edMsg);
        assert Ed25519Example.verify(edKeys.getPublic(), edMsg, edSig);

        // X25519-HKDF
        KeyPair alice = X25519HkdfExample.generateKeyPair();
        KeyPair bob = X25519HkdfExample.generateKeyPair();
        byte[] salt = CryptoSupport.randomBytes(16);
        byte[] info = "info".getBytes(StandardCharsets.UTF_8);
        byte[] aliceKey = X25519HkdfExample.hkdfSha256(X25519HkdfExample.sharedSecret(alice, bob.getPublic()), salt, info, 32);
        byte[] bobKey = X25519HkdfExample.hkdfSha256(X25519HkdfExample.sharedSecret(bob, alice.getPublic()), salt, info, 32);
        assert CryptoSupport.constantTimeEquals(aliceKey, bobKey);

        // AES-KeyWrap
        javax.crypto.SecretKey kek = AesKeyWrapExample.generateAesKey();
        javax.crypto.SecretKey contentKey = AesKeyWrapExample.generateAesKey();
        byte[] wrapped = AesKeyWrapExample.wrap(kek, contentKey);
        javax.crypto.SecretKey recovered = AesKeyWrapExample.unwrap(kek, wrapped);
        assert Arrays.equals(contentKey.getEncoded(), recovered.getEncoded());

        System.out.println("All Java tests passed!");
    }
}
